inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/detection_output_gpu_ref.cl

```diff
@@ -0,0 +1,729 @@
+    }
+    return final_detections;
+}
+
+UNIT_TYPE FUNC(jaccardOverlap)(UNIT_TYPE* bbox1, UNIT_TYPE* bbox2)
+{
+    UNIT_TYPE overlap = 0.0;
+    bool intersecting = (bbox1[0] < bbox2[2]) & (bbox2[0] < bbox1[2]) & (bbox1[1] < bbox2[3]) & (bbox2[1] < bbox1[3]);
+
+    if (intersecting)
+    {
+        const UNIT_TYPE intersect_width = min(bbox1[2], bbox2[2]) - max(bbox1[0], bbox2[0]);
+        const UNIT_TYPE intersect_height = min(bbox1[3], bbox2[3]) - max(bbox1[1], bbox2[1]);
+        if (intersect_width > 0 && intersect_height > 0) {
+            const UNIT_TYPE intersect_size = intersect_width * intersect_height;
+            const UNIT_TYPE bbox1_size = (bbox1[2] - bbox1[0]) * (bbox1[3]- bbox1[1]);
+            const UNIT_TYPE bbox2_size = (bbox2[2] - bbox2[0]) * (bbox2[3]- bbox2[1]);
+            overlap = intersect_size / (bbox1_size + bbox2_size - intersect_size);
+        }
+    }
+    return overlap;
+}
+
+UNIT_TYPE FUNC(get_confidence_offset)(const uint idx_prior, const uint idx_class, const uint idx_image)
+{
+    return (idx_prior * NUM_CLASSES + idx_image * NUM_OF_PRIORS * NUM_CLASSES + idx_class) * CONF_XY_SIZE_PRODUCT + CONF_PADDING;
+}
+
+UNIT_TYPE FUNC(get_largest_score)(__global UNIT_TYPE* input_confidence, const uint idx_prior, const uint idx_image)
+{
+    const uint idx_start = (BACKGROUND_LABEL_ID == 0 ? 1 : 0);
+    uint offset = FUNC_CALL(get_confidence_offset)(idx_prior, idx_start, idx_image);
+    UNIT_TYPE max_score = input_confidence[offset];
+    int idx = idx_start;
+
+    for (uint j = idx_start; j < NUM_CLASSES; j++)
+    {
+        offset = FUNC_CALL(get_confidence_offset)(idx_prior, j, idx_image);
+        UNIT_TYPE score = input_confidence[offset];
+        if (score > max_score) {
+            max_score = score;
+            idx = j;
+        }
+    }
+    return idx;
+}
+
+#ifdef DO_STAGE_0_CAFFE
+KERNEL (detection_output_ref_stage_0_caffe)(
+    __global UNIT_TYPE* input_confidence,
+    __global uchar *buffer0,
+    __global int *buffer2)
+{
+    const int batchId = get_global_id(0);
+    const int classId = get_global_id(1);
+    const int box_gid = get_global_id(2);
+
+    const int start_bid = box_gid * NUM_SCORE_PER_ITEM;
+    const int end_bid = min(start_bid + NUM_SCORE_PER_ITEM, NUM_OF_PRIORS);
+
+    __local char bit_mask[NUM_BIT_MASK];
+    __local int block_num[NUM_SCORE_BLOCK];
+
+    block_num[box_gid] = 0;
+
+    {
+        int mask_id = start_bid / 8;
+        int total_block_selected_num = 0;
+        for (int i = start_bid; i < end_bid; i += 8) {
+            char mask = 0;
+            for (int bi = 0; bi < 8; bi++) {
+                if ((i + bi) >= NUM_OF_PRIORS)
+                    break;
+
+                UNIT_TYPE score = FUNC_CALL(get_score)(input_confidence, (i + bi), classId, batchId);
+                if (score == -1)
+                    continue;
+
+                mask |= (1 << bi);
+                total_block_selected_num++;
+            }
```

Comment:
A suggestion:
`unroll_for(int bi = 0; ...) {
    if((i + bi) >= NUM_OF_..) break;
    UNIT_TYPE score = FUNC_CALL(get_score(...)....;
     int has_score = (score == -1) ? 0 : 1;
     mask |= (has_score << bi);
     total_block_selected_num += has_score; 
 }
`
Just to reduce a branch.

Reply:
and also change get_score as inline function..

Reply:
Fixed as your suggestion